### PR TITLE
API: Register existing tables in Iceberg HiveCatalog

### PR DESCRIPTION
**Pull request by anuragmantri**
This PR allows us to register existing tables in Iceberg HiveCatalog.

Right now, we can keep metadata and data files while dropping tables from Iceberg HiveCatalog. However, there is no way to register those tables back even though the metadata and data files can be still there. So, we need to extend HiveCatalog with another method that will accept a location to the valid metadata file and would register a table in HMS.

This will allow us to properly support external tables in Spark.

### CODE REVIEW

File: hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java [hunk 1]
```diff
@@ -370,6 +371,85 @@ public void testNonDefaultDatabaseLocation() throws IOException, TException {
...
+
+    Map<String, String> newTableParameters = newTable.getParameters();
+    Assert.assertNull(newTableParameters.get(PREVIOUS_METADATA_LOCATION_PROP));
+    Assert.assertEquals(originalParams.get(TABLE_TYPE_PROP), newTableParameters.get(TABLE_TYPE_PROP));
+    Assert.assertEquals(originalParams.get(METADATA_LOCATION_PROP), newTableParameters.get(METADATA_LOCATION_PROP));
+    Assert.assertEquals(originalTable.getSd(), newTable.getSd());
+  }
+
+  @Test
+  public void testCloneTable() throws IOException {
```
Comment (RussellSpitzer) [inline hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java:401]: I would remove this test just because I don't want to give anyone ideas :)

Comment (anuragmantri) [reply hive-metastore/src/test/java/org/apache/iceberg/hive/HiveTableTest.java]: Agreed :) Removed this test.

**Comment by pvary** [comment]
Can I use this to register an arbitrary table from any catalog (HadoopTable or HadoopCatalog or GlueCatalog) to the HiveCatalog?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 2]
```diff
@@ -211,6 +214,18 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
     }
   }
 
+  @Override
+  public org.apache.iceberg.Table registerTable(TableIdentifier identifier, String metadataFileLocation) {
+    Preconditions.checkArgument(isValidIdentifier(identifier), "Invalid identifier: %s", identifier);
+
+    TableOperations ops = newTableOps(identifier);
+    HadoopInputFile metadataFile = HadoopInputFile.fromLocation(metadataFileLocation, conf);
```
Comment (szehon-ho) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:222]: Shouldn't we use FileIO to get the InputFile instead of hardcoding HadoopFileIO (in case we are using S3FileIO, for example?)

File: api/src/main/java/org/apache/iceberg/catalog/Catalog.java [hunk 3]
```diff
@@ -340,6 +340,17 @@ default boolean dropTable(TableIdentifier identifier) {
    */
   Table loadTable(TableIdentifier identifier);
 
+  /**
+   * Register a table.
```
Comment (szehon-ho) [inline api/src/main/java/org/apache/iceberg/catalog/Catalog.java:344]: Seems from the test, the table is not there and it will create one from the file.  It wasn't too obvious from the name of the API, can we enhance the javadoc a bit to detail that?

**Review by szehon-ho — CHANGES_REQUESTED** [review]
Logic looks good to me, had some comments

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 2 above)
Comment (flyrain) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:222]: Agree with Szehon. `fileIO` should be used here.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java [hunk 4]
```diff
@@ -210,7 +210,8 @@ protected void doRefresh() {
   @SuppressWarnings("checkstyle:CyclomaticComplexity")
   @Override
   protected void doCommit(TableMetadata base, TableMetadata metadata) {
-    String newMetadataLocation = writeNewMetadata(metadata, currentVersion() + 1);
+    String newMetadataLocation = base == null && metadata.metadataFileLocation() != null ?
+        metadata.metadataFileLocation() : writeNewMetadata(metadata, currentVersion() + 1);
```
Comment (flyrain) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:214]: Looks like this change is not related. Can we remove it?

Comment (flyrain) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:214]: Is it used by the new test?

File: api/src/main/java/org/apache/iceberg/catalog/Catalog.java (hunk 3 above)
Comment (flyrain) [reply api/src/main/java/org/apache/iceberg/catalog/Catalog.java:344]: What happens if the table is already there? Do we throw an exception?

Comment (anuragmantri) [reply api/src/main/java/org/apache/iceberg/catalog/Catalog.java]: Updated the java doc to mention this API will register a table with the catalog if it does not exist. It throws an exception if it exists. I have also added a unit test case.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 2 above)
Comment (anuragmantri) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java]: Thanks for catching this. Updated to use `FileIO`.

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java (hunk 4 above)
Comment (anuragmantri) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveTableOperations.java:214]: Without this change, registering a table creates a new metadata file with a new version instead of using the version provided by the metadata file. Yes, the tests also rely on this.

**Comment by anuragmantri** [comment]
> Can I use this to register an arbitrary table from any catalog (HadoopTable or HadoopCatalog or GlueCatalog) to the HiveCatalog?

@pvary - Yes, all this needs is the latest metadata file.

**Comment by pvary** [comment]
> @pvary - Yes, all this needs is the latest metadata file.

This was my conclusion too (after I have left the comment I continued chewing on this one 😄) Then I went further, and this is where I am now:
- What happens, if a table is modified in one of the catalogs (I suspect that this is catalog dependent as HiveCatalog will not follow the changes, but HadoopTable/HadoopCatalog might follow each others changes, as they rely on the file names) 
- Also issues will arise if we start expiring snapshots. If one catalog decides that one of the files are not needed anymore, then it could remove the file, even if the file is still needed for the other table. 

Is there a way to prevent this situation? 

Thanks, Peter

File: api/src/main/java/org/apache/iceberg/catalog/Catalog.java [hunk 5]
```diff
@@ -340,6 +340,18 @@ default boolean dropTable(TableIdentifier identifier) {
    */
   Table loadTable(TableIdentifier identifier);
 
+  /**
+   * Register a table with the catalog if it does not exist.
```
Comment (rdblue) [inline api/src/main/java/org/apache/iceberg/catalog/Catalog.java:344]: FYI @bryanck. What do you think about this?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java [hunk 6]
```diff
@@ -211,6 +214,23 @@ public void renameTable(TableIdentifier from, TableIdentifier originalTo) {
     }
   }
 
+  @Override
+  public org.apache.iceberg.Table registerTable(TableIdentifier identifier, String metadataFileLocation) {
```
Comment (bryanck) [inline hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:218]: Could we put this in BaseMetastoreCatalog? It doesn't look like there is anything Hive-specific here, so other catalog implementations could potentially benefit.

File: api/src/main/java/org/apache/iceberg/catalog/Catalog.java (hunk 5 above)
Comment (bryanck) [reply api/src/main/java/org/apache/iceberg/catalog/Catalog.java:344]: LGTM, it is very straightforward, I just had one question below...

**Comment by anuragmantri** [comment]
> * What happens, if a table is modified in one of the catalogs (I suspect that this is catalog dependent as HiveCatalog will not follow the changes, but HadoopTable/HadoopCatalog might follow each others changes, as they rely on the file names)
> * Also issues will arise if we start expiring snapshots. If one catalog decides that one of the files are not needed anymore, then it could remove the file, even if the file is still needed for the other table.
> 

Thanks for the questions @pvary. I haven't really thought about concurrent access to the tables from multiple catalogs. It did come up a little bit in our [DR discussion](https://lists.apache.org/thread/8cg14cvs0hxybpvz20olr78ysm7xrtw2) in the dev list. I'm not sure this should be allowed. 

@aokolnychyi - What are your thoughts on this?

File: hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java (hunk 6 above)
Comment (anuragmantri) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:218]: This is a good point. Potentially, other catalogs can benefit from this. However, `FileIO` is initialized with the catalog and there maybe custom implementations passed as catalog properties. I'm not sure yet on how to move this logic to `BaseMetastoreCatalog`. Do you mind if I do that as a seperate enhancement PR to this one?

Comment (bryanck) [reply hive-metastore/src/main/java/org/apache/iceberg/hive/HiveCatalog.java:218]: Sure, thanks

**Review by rdblue — APPROVED** [review]

**Comment by rdblue** [comment]
This looks good to me when tests are passing.

**Comment by anuragmantri** [comment]
Looks like the build failed. Is there a way to re-trigger the build without pushing the change?

**Comment by flyrain** [comment]
Looks like this build errors happened in multiple PRs. I saw it #3745 as well. It could be an issue in build infra or some changes in Gradle. @rdblue 
```
A problem occurred configuring root project 'iceberg'.
31
> Could not resolve all files for configuration ':classpath'.
32
   > Could not resolve me.champeau.jmh:jmh-gradle-plugin:0.6.6.
33
     Required by:
34
         project :
35
      > Could not resolve me.champeau.jmh:jmh-gradle-plugin:0.6.6.
36
         > Could not get resource 'https://plugins.gradle.org/m2/me/champeau/jmh/jmh-gradle-plugin/0.6.6/jmh-gradle-plugin-0.6.6.module'.
37
            > Could not GET 'https://jcenter.bintray.com/me/champeau/jmh/jmh-gradle-plugin/0.6.6/jmh-gradle-plugin-0.6.6.module'. Received status code 502 from server: Bad Gateway
38
   > Could not resolve org.apache.logging.log4j:log4j-core:2.17.0.
```

**Comment by bryanck** [comment]
Seems like jcenter is having issues... https://status.gradle.com

**Comment by flyrain** [comment]
> Seems like jcenter is having issues... https://status.gradle.com

The Jcenter is back online now. The build should pass once it is retriggered.

**Comment by anuragmantri** [comment]
Retest this please

**Review by RussellSpitzer — APPROVED** [review]
No idea why the Javadoc build is breaking, I re ran it several times through the workflow. But the links it can't get the resource from also don't work for me.

**Comment by RussellSpitzer** [comment]
Tests Passed and Merged! Thanks @anuragmantri and all reviewers!

**Comment by anuragmantri** [comment]
Thanks to @aokolnychyi, the original author of the PR and everyone for the review!
